### PR TITLE
 feat: add image support on ecr repos

### PR DESCRIPTION
**Pull request by yrobla**
Signed-off-by: Yolanda Robla <info@ysoft.biz>
Closes: #1230

### CODE REVIEW

File: src/modules/0.0.20/aws_ecr/index.ts [hunk 1]
```diff
@@ -9,7 +18,182 @@ import { policiesAreSame } from '../../../services/aws-diff';
...
+      const opts = repositoryName
+        ? {
+            where: {
+              repositoryName,
+            },
+          }
+        : {};
+      return await ctx.orm.find(RepositoryImage, opts);
+    },
+  });
```
Comment (aguillenv) [inline src/modules/0.0.20/aws_ecr/index.ts:113]: why do we need this special db read method?

File: src/modules/0.0.20/aws_ecr/index.ts [hunk 2]
```diff
@@ -9,7 +18,182 @@ import { policiesAreSame } from '../../../services/aws-diff';
...
+    },
+    updateOrReplace: () => 'update',
+    update: async (es: RepositoryImage[], ctx: Context) => {
+      return [];
+    },
+    delete: async (es: RepositoryImage[], ctx: Context) => {
+      const client = (await ctx.getAwsClient()) as AWS;
+      for (const e of es) {
+        if (e.privateRepository) await this.deleteRepositoryImage(client.ecrClient, e.privateRepository);
+        else if (e.publicRepository) await this.deleteRepositoryImage(client.ecrClient, e.publicRepository);
```
Comment (aguillenv) [inline src/modules/0.0.20/aws_ecr/index.ts:186]: I think we are using the wrong client here for delete the public images

**Review by dfellis — APPROVED** [review]

File: src/modules/0.0.20/aws_ecr/index.ts [hunk 3]
```diff
@@ -105,6 +346,8 @@ class PublicRepositoryMapper extends MapperBase<PublicRepository> {
     delete: async (es: PublicRepository[], ctx: Context) => {
       const client = (await ctx.getAwsClient()) as AWS;
       for (const e of es) {
+        // need to delete images associated with this repository
+        await this.module.repositoryImages.deletePublicRepositoryImages(client.ecrPubClient, e);
```
Comment (depombo) [inline src/modules/0.0.20/aws_ecr/index.ts:350]: hmm I am not 100% sure about this one. we would be overriding the database functionality. The user should have deleted entries in the `image` table or done a delete query on `public_repository` with `CASCADE` and then the lazyLoader should converge on deleting the images and then the repository?

File: src/modules/0.0.20/aws_ecr/index.ts [hunk 4]
```diff
@@ -256,6 +499,9 @@ class RepositoryMapper extends MapperBase<Repository> {
     delete: async (es: Repository[], ctx: Context) => {
       const client = (await ctx.getAwsClient()) as AWS;
       for (const e of es) {
+        // need to delete images associated with this repository
+        await this.module.repositoryImages.deleteRepositoryImages(client.ecrClient, e);
+
```
Comment (depombo) [inline src/modules/0.0.20/aws_ecr/index.ts:504]: same comment as above

File: test/modules/aws-ecr-integration.ts [hunk 5]
```diff
@@ -1,251 +1,475 @@
...
+    query(
+      `
     INSERT INTO aws_credentials (access_key_id, secret_access_key)
     VALUES ('${process.env.AWS_ACCESS_KEY_ID}', '${process.env.AWS_SECRET_ACCESS_KEY}')
-  `, undefined, false));
+  `,
+      undefined,
+      false,
+    ),
+  );
```
Comment (depombo) [inline test/modules/aws-ecr-integration.ts:60]: the format seems to be changing drastically which is totally fine, but there are so many re-formats that it makes the *actual* changes harder to track. can we avoid this in the future by doing it in separate PRs / commits?

Comment (depombo) [reply test/modules/aws-ecr-integration.ts:60]: (this is a nit)

**Comment by yrobla** [comment]
@depombo i tried to use the db approach you suggested. But i found a problem with on_cascade property. If i setup in manytoone side as i did there, it is not deleting the related images.
If i add the onDelete:cascade on the onetomany side, it throws an error: update or delete on table "repository" violates foreign key constraint "FK_3c7183f97647067d207d9276df6" on table "repository_image"

What we need here is to actually delete the associated images but before deleting the parent row. What should be the right settings to setup in the models to achieve it?

**Comment by depombo** [comment]
> @depombo i tried to use the db approach you suggested. But i found a problem with on_cascade property. If i setup in manytoone side as i did there, it is not deleting the related images. If i add the onDelete:cascade on the onetomany side, it throws an error: update or delete on table "repository" violates foreign key constraint "FK_3c7183f97647067d207d9276df6" on table "repository_image"
> 
> What we need here is to actually delete the associated images but before deleting the parent row. What should be the right settings to setup in the models to achieve it?

Mmm I would take a look at what we do in ECS, maybe we need this https://github.com/iasql/iasql-engine/blob/main/src/modules/0.0.20/aws_ecs_fargate/entity/task_definition.ts#L113. I am not sure. We could update the remaining fk constraints that get violated to null as described here
https://stackoverflow.com/questions/29305653/postgresql-deleting-rows-referenced-from-another-table, but I am actually not sure how to express that in typeorm. Tbh I would leave the models alone, delete the lines I initially commented on and explicitly delete images under a given repository in the tests.

**Comment by yrobla** [comment]
I tried your cascade approach but it actually tries to first delete the repo and then the images. And it cannot be done because aws complains that cannot delete a repo with images.
So you prefer that we do not remove images automatically when we delete the repo but instead user has to delete that manually ?

File: test/modules/aws-ecr-integration.ts [hunk 6]
```diff
@@ -1,251 +1,494 @@
...
-  
-    it('deletes the public repository', query(`
+
+    it(
+      'deletes the repository images',
+      query(`
+    DELETE FROM repository_image WHERE public_repository= '${pubRepositoryName}';
+  `),
+    );
+    it('applies deletes the public repository image', apply());
```
Comment (depombo) [inline test/modules/aws-ecr-integration.ts:427]: this should work without this `apply` or this here on purpose? the lazyLoader should figure out the correct order to delete the resources in even if there is a transient error

**Review by depombo — APPROVED** [review]

**Pull request merged**